Clean estimated histogram

amnonh opened the pull request:
These series clean the estimated histogram implementation by removing FIXME and if0 from the code and allow finer and customized histogram report.

It uses that implementation for the read and write latency histogram report.

amnonh commented:
I would like to add additional functionality to estimated_histogram, it will be best if we can merge the clean up patch first

slivne commented:
@glommer can you please bless so I can get avi to review and merge

glommer commented:
It works as a cleanup patch, and I see that amnon reimplements min() later.
From my PoV this is okay, Avi may want a different patch split

amnonh commented on utils/estimated_histogram.hh:
```diff
@@ -247,66 +238,12 @@ public:
-    /**
-     * @return the smallest value that could have been added to this histogram
-     */
-    public long min()
```
it's implemented, but we didn't remove the java code, so I did

slivne commented:
@avikivity ping

slivne commented:
@avikivity ping we have globbers blessing on this

avikivity commented on utils/estimated_histogram.hh:
```diff
@@ -339,16 +276,7 @@ public:
-        }
-
-        return ((double) (sum + elements -1)/ elements);
+        return (_count) ? ((double)_sample_sum)/_count : 0;
```
Yes, @amnonh please fix.

avikivity commented on utils/estimated_histogram.hh:
```diff
@@ -73,13 +73,59 @@ struct estimated_histogram {
+     * on the histogram usage and return a minimized histogram
+     */
+    template<class T>
+    seastar::metrics::histogram get_histogram(const std::vector<T>& buckets_limit) const {
```
Why not pass base, ratio, and count? It looks too generic.

Also, this shouldn't be a template is there are just two possible Ts. You can have a private implenentation for two non-template public funcitons.

amnonh replied:
the incentive for the change is that exponential is good for lower values, but not so good for higher values.

I'll change to non-template implmentation

avikivity commented on utils/estimated_histogram.hh:
```diff
@@ -352,6 +388,43 @@ public:
+    static const std::vector<duration>& get_duration_vector() {
+        static std::vector<duration> res {
+            std::chrono::microseconds(16),
+            std::chrono::microseconds(32),
```
Where does this come from?

avikivity commented on service/storage_proxy.cc:
```diff
@@ -809,8 +809,8 @@ storage_proxy::storage_proxy(distributed<database>& db, storage_proxy::config cf
     , _view_update_handlers_list(std::make_unique<view_update_handlers_list>()) {
     namespace sm = seastar::metrics;
```
The distribution above is exponential, so the patch description is incorrect.

amnonh replied:
not entirely, it replaces the exponential distribution we had with something more forgiven near the 1s buckets

amnonh replied:
right, but everything on the way supports it, we can get finer resolution when we need it

amnonh replied:
in that case we can keep the 1.2 distribution that is currently reported, we want to lower the number of buckets from the number that is used to collect the data (90) to a lower number

amnonh commented:
@avikivity I've forced push a new version with the changes that are not addressed in the review

amnonh commented:
@avikivity I have forced push v3, I use a 1.3 ratio, it uses a bit more buckets but I guess it's not the end of the world.

We should expect lower latencies following this patch

slivne commented:
@amnonh lease rebase so we can ping avi to rereview

amnonh commented:
I have pull rebase to master

slivne commented:
@avikivity ping

slivne commented:
@avikivity ping

slivne commented:
@avikivity ping

bhalevy commented:
target branch should be `next`
and in addition to this patch, need to fix https://github.com/scylladb/scylla/issues/5577

slivne commented:
@avikivity whats blocking this fix - we also have a customer issue that needs to be fixed ontop of this.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -1382,10 +1382,10 @@ storage_proxy::storage_proxy(distributed<database>& db, storage_proxy::config cf
+        sm::make_histogram("read_latency", sm::description("The general read latency histogram"), [this]{ return _stats.estimated_read.get_histogram(std::chrono::microseconds(500), 33, 1.3);}),
+        sm::make_histogram("write_latency", sm::description("The general write latency histogram"), [this]{return _stats.estimated_write.get_histogram(std::chrono::microseconds(500), 33, 1.3);}),
+        sm::make_histogram("cas_read_latency", sm::description("Transactional read latency histogram"), [this]{ return _stats.estimated_cas_read.get_histogram(std::chrono::microseconds(500), 33, 1.3);}),
+        sm::make_histogram("cas_write_latency", sm::description("Transactional write latency histogram"), [this]{return _stats.estimated_cas_write.get_histogram(std::chrono::microseconds(500), 33, 1.3);}),
```
I don't see a 3-argument version of get_historgram. Did the patches get reordered by github?

amnonh replied:
yes

avikivity commented on utils/estimated_histogram.hh:
```diff
@@ -73,40 +73,59 @@ struct estimated_histogram {
+     * and what is the ratio between bucket limits.
+     * Setting the bucket_ratio for 1 or less, will use the buckets as-is
+     */
+    seastar::metrics::histogram get_histogram(size_t lower_bucket = 1, size_t max_buckets = 20, double bucket_ratio = 1.44) const {
```
I don't understand how 1.3 gives more accurate results than the original 1.2.

avikivity replied:
And why we use 1.44 for the default

amnonh replied:
We never use the original 1.2, up until now we use a factor of 2, which appears to be too high.
I pick sqrt(2) it seems to make sense, I can pick anything else including the maximum available.

avikivity commented on utils/estimated_histogram.hh:
```diff
@@ -73,40 +73,59 @@ struct estimated_histogram {
+                last_bound = bucket_offsets[pos];
+            } else {
+                last_bound *= bucket_ratio;
+            }
```
I don't understand this bucket_ratio < 1 thing. Isn't there a bucket_ratio that corresponds to the current code? If so, it can be the default.

amnonh replied:
It eliminates all the ratio calculation, so it allows the caller to say: give me the maximal precision available.
If we know that we want the maximal precision, why do the calculation? instead, just take the value.

avikivity replied:
I don't understand what's the maximal precision, since we create buckets in any case.

amnonh replied:
We start with the base estimated histogram (typically, uses a ratio of 1.2 and 92 bucket but can use anything else).
When we report it, we use an accumulative histogram with an (optional) different ratio (i.e. different precision).
So the user of the accumulative histogram (Prometheus) can choose to limit the size of the histogram it will get, or ask for the highest precision histogram available.
The idea is to allow requesting the histogram without reducing the number of buckets.

A valid alternative is to report it by default.
If we are reporting all the buckets, there is no point in calculating the buckets limits, instead, we can just copy the values.

avikivity replied:
I still don't understand what bucket_ratio <= 1 means. Is it a way to say "change the method of calculating ratios to something else"? If so it is a bad method because no one will be able to guess it.

avikivity replied:
I see there's a comment that says so. But it's a bad API. Better is to use a function with a different name (internally they can share implementations).

amnonh replied:
I'll replace it all anyhow, this is a step that keeps the current API.

avikivity commented on utils/estimated_histogram.hh:
```diff
@@ -360,8 +360,6 @@ public:
         return s;
     }
-
-    }
```
There is another, more efficient way to do this, and that is by using pseudo floating point. For example we can have 4 significant bits and then a value of 195 = 0xc3 = 0y1100'0011 = (128) * (1 + (0y100'0011 / 128)) =~ 128 * (1 + 0y100 / 8) = (1<<7)*(1 + 4/8)

The bucket number is therefore 7*8+4 = 60. To recover the value again divide by 8 (we get 7 with a reminder of 4), and write (1<<7) + (4 << (7 - 3)) = 128 + 64 = 192.

amnonh replied:
I'll have a look

amnonh replied:
I've openned https://github.com/scylladb/scylla/issues/5815

slivne commented:
@amnonh ping

amnonh commented:
@avikivity I have forced push a new version (also cleared the conflicts)
I have addressed your comments here, but if there is something else you want to change I can do it.

Main changes, the default generated histogram now uses 1.3 ratio and reduce the number of buckets to 33 (about 1/3 of the original histogram)
I've based it on your PR comment that looks reasonable.
Additional changes to the histogram will be based on the cleaner version

slivne commented:
@avikivity can you please review this

slivne commented:
@avikivity ping

amnonh commented:
@avikivity I would love to move to the next step (after this initial clean up is completed) for an optimized histogram that will be widely used.
Is there's anything else that stops this PR from getting in?

slivne commented:
@avikivity ping

avikivity commented on utils/estimated_histogram.hh:
```diff
@@ -73,40 +73,58 @@ struct estimated_histogram {
 
     int64_t _count = 0;
```
This isn't "Clean estimated histogram".

amnonh replied:
Are you talking about the PR header?
The patch is about getting better estimation

avikivity replied:
Yes. The subject should summarize what the patches are doing.

avikivity commented on utils/estimated_histogram.hh:
```diff
@@ -73,40 +73,58 @@ struct estimated_histogram {
 
-    seastar::metrics::histogram get_histogram(size_t lower_bucket = 1, size_t max_buckets = 16) const {
+    /*
+     * \brief get an accumulative histogram
```
s/accumulative/cumulative/

avikivity commented on utils/estimated_histogram.hh:
```diff
@@ -73,40 +73,58 @@ struct estimated_histogram {
+     * what is the number of buckets and what is the ratio between bucket limits.
+     * Setting the bucket_ratio to 1 or less, will return all the buckets.
+     */
+    seastar::metrics::histogram get_histogram(size_t lower_bucket = 1, size_t max_buckets = 33, double bucket_ratio = 1.3) const {
```
Why the change to 33? Not that I think there should be a default here (except maybe all).

avikivity commented on utils/estimated_histogram.hh:
```diff
@@ -174,6 +192,12 @@ public:
+        }
+        if (!_max || n > _max) {
+            _max = n;
+        }
```
Please use min_tracker/max_tracker. Using 0 as a flag value is dangerous.

avikivity commented on utils/estimated_histogram.hh:
```diff
@@ -192,31 +216,15 @@ public:
-            i++;
-        }
-        return 0;
+        return _min;
```
What's the motivation? we get better min/max, but also do more work per operation. Is there a problem with the current min/max?

amnonh replied:
it's not a critical problem, but there is no other way of knowing what is the real min and max.

avikivity commented on utils/estimated_histogram.hh:
```diff
@@ -233,6 +241,8 @@ public:
         _count += b._count;
         _sample_sum += b._sample_sum;
+        _min = (b._min < _min) ? b._min : _min;
+        _max = (b._max > _max) ? b._max : _max;
```
std::min/std::max.

What if one of them had _min = 0 because there were no values?

avikivity commented on utils/estimated_histogram.hh:
```diff
@@ -276,16 +286,7 @@ public:
-        }
-
-        return ((double) (sum + elements -1)/ elements);
+        return (_count) ? ((double)_sample_sum)/_count : 0;
```
Again we get better accuracy, but at the cost of doing more work. It's negligible compared to the work done by searching for the right bucket (which will be eliminated by using fixed-point pseudo-logarithms), but still should be justified.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -1349,7 +1349,7 @@ void storage_proxy_stats::write_stats::register_stats() {
     _metrics.add_group(COORDINATOR_STATS_CATEGORY, {
```
Please add the word "change" to the subject so it becomes a sentence.

avikivity replied:
The proemtheus text format is so :facepalm:

avikivity commented on service/storage_proxy.cc:
```diff
@@ -1444,7 +1444,7 @@ void storage_proxy_stats::stats::register_stats() {
         sm::make_histogram("read_latency", sm::description("The general read latency histogram"),
                 {storage_proxy_stats::current_scheduling_group_label()},
-                [this]{ return estimated_read.get_histogram(16, 20);}),
+                [this]{ return estimated_read.get_histogram(std::chrono::microseconds(500), 33, 1.3);}),
```
I think we need a few more buckets, so a 4-second query isn't shown as +inf.

avikivity replied:
maybe a better ratio is sqrt(2) or cuberoot(2), so the powers of two are still represented. It's not really important, but it's nicer to see numbers such as 1ms or 16ms. Really a cosmetic issue.

amnonh replied:
I'm going to drop all that ratio as soon as I'll move to the new implementation based on your comments on the previous review.

amnonh commented:
@avikivity without the min/max optimization and sense it will not go into 4.1, there is no point to add the part of a better resolution and I'll delay it to a different PR based on pseudo floating point.

This lives what I hope is noncontroversial, the cleaning part of the patch.

This is a pull rebase of that part

avikivity commented:
Queued as 515c77e881
